### ISSUE TITLE
Support for Texture (AsyncDisplayKit)

### ISSUE THREAD

**Issue by zboralski** [lvnkmn/Zoomy#70]
It would be great if [Texture](https://github.com/TextureGroup/Texture) (AsyncDisplayKit) was supported. 

ASImageNode are pretty much equivalent to a UIImageView. 

I am gonna try to populate some dummy UIImageView as my nodes enter display range.

**Comment by lvnkmn** [comment]
Hi @zboralski,

It's been a while that I've used async displaykit, which was in the objc days. But I think this might just be possible. 

Can you tell me if imagenodes work with gesture recognizers when added to their `.view`?

If so, I think this could work when Zoomy starts relying on types of an newly introduced `Zoomable` protocol.  `Imagenodes` could then conform to this using an extension which could be added within a subspec for Texture.  `UIImageViews` will also conform to this. 

After you've answered my question, I think I should have time for the `Zoomable` protocol part. I'll create a branch for you to test it in.

**Comment by zboralski** [comment]
Oh that'll be amazing!

All classes derived from ASDisplayNodes work with gesture recognisers, ASNetworkImageNode, ASImageNode, etc, you just apply the gesture recognisers to the underlying view:

for example:

```swift
imageNode.onDidLoad({ node in
  node.view.addGestureRecognizer(tap)
}

Happy to help test and contribute.

**Comment by lvnkmn** [comment]
That looks promising ☝️

So far I've done most (if not all) of the needed implementations on the Lib's side.

With a bit of luck I'll have time to create a small example which uses ImageNodes later today, otherwise tomorrow it should definitely be ready for testing.

I'll let you know when the branch is ready.

**Comment by lvnkmn** [comment]
Allright, as you can follow in the commits above, I've created an example in which Texture works using the instazoom settings.

With the current state of the code there's a couple of gotcha's 

- It will only work when zoomed in the ImageNodes views superview.
  - This superview is only available after viewDidAppear (maybe viewWillAppear, didn't test yet)
  - By default the viewControllers view is used, but in this case the absolute coordinates that are needed by Zoomy cannot be calculated.

For non insta style zooming (feel free to try), some more work is needed, I suspect in `absoluteFrame(of subjectView: UIView?)` of `ImageZoomController`

I'll have a look at it again tomorrow. For now, feel free to check all of the above out on: https://github.com/lvnkmn/Zoomy/tree/feature/%2370_-_support_for_texture

**Comment by zboralski** [comment]
That's good I need an instazoom functionality :) Thanks. I'll test it now.

**Comment by zboralski** [comment]
Great I tested the example and it is really nice. 

I did a quick unsuccessful test in my code and I think I need to solve the conflict with my existing gesture recognisers. I have to look if the easiest solution is to make them fail if two fingers are pressed.

What's the setup cost of these zoomables? I can add the behaviour when they enter visible range and tear them down when they exit prefetch.

**Comment by zboralski** [comment]
I disabled my gesture recognisers and added some logging. My imageNode don't conform to zoomable... let me take a look

**Comment by zboralski** [comment]
Ah I missed that extension line at the bottom :) it works!

**Comment by zboralski** [comment]
Amazing! 

Now I have to figure out how to make it appear outside of my cell node frame. All my nodes are inside a ASCollectionNodeController. I'll check in the morning.

**Comment by lvnkmn** [comment]
A quick and dirty way to achieve this, is by setting your cell nodes views `clipsToBounds` property to false. 

The way it normally works in Zoomy is adding the zoom behavior **in** a view where the zoomable view is a descendant of. When it’s not, it can currently not calculate an absolute frame inside the container view. 

I suspect the proper way to solve this is improving the absolute frame calculation, which might just solve the non Insta zoom behaviors as well.

**Comment by lvnkmn** [comment]
While you are at it with testing, can you maybe update the example to use `ASCollectionNodeConTroller` since my Texture skillls are a bit rusty?

**Comment by zboralski** [comment]
Sure, I'll create an example for ASCollectionNodeController. I am traveling this weekend. I'll find some time.

**Comment by zboralski** [comment]
clip to bound only work above the view but not below...

Is there a way to allow only allow zooming in but not zooming out (shrinking below the original size of the view?

**Comment by zboralski** [comment]
<img width="487" alt="image" src="https://user-images.githubusercontent.com/104316/64419462-de9bcb00-d094-11e9-89fc-13c2d47d17bb.png">

**Comment by zboralski** [comment]
what's your email I'll send you a TestFlight invite? you can message me on Twitter, same username.

**Comment by lvnkmn** [comment]
Sorry didn’t have time for this lib today. There’s currently no minimum zoomscale in the settings, but should be relatively easy to add. I have time to dig into this again on Monday.  Enjoy your weekend!

**Comment by lvnkmn** [comment]
Something else that could help you right now is hiding views as soon between the will begin and end displaying overlay methods that are called when the viewcontroller conforms to Zoomy.Delegate.

**Comment by lvnkmn** [comment]
Hi @zboralski could you still create that ASCollectionNode example? That way it will be a lot easier to debug what's going on.

**Comment by lvnkmn** [comment]
In the mean time I've found that ASNodes don't really respond well to their alpha being set to 0, normally views don't turn black when doing this. This is something I will also look into when the new example is there.

**Comment by lvnkmn** [comment]
Ok, @zboralski the latest commit on this branch showed me everything now works like it does using UIImageViews. Can you confirm this? If so I'll start creating a subspec (containing the needed extension) and start shipping this feature.

**Comment by lvnkmn** [comment]
No worries about the delay. Curious to know though if you'd still have time for this. If you don't, I'll schedule some time to do it myself. 

Also I'd like to know what your current issues are. I just added support for Swift Package manager and I'd love to release this feature soon.

**Comment by lvnkmn** [comment]
Allright, I've added a collectionView example and for me it works 100% with both insta an scrollable zooming. 

As soon as I've found the right way to deal with the needed `extension ASImageNode: Zoomable {}` I'll create a release for this.

**Comment by lvnkmn** [comment]
Done and released, it's now part of [4.0.0](https://github.com/lvnkmn/Zoomy/releases/tag/4.0.0)
